contracts/ics/chan/src/entry.rs

```diff
@@ -0,0 +1,125 @@
+use crate::error::{Error, Result};
+use alloc::vec::Vec;
+use ckb_ics_axon::handler::{
+    handle_channel_open_ack_and_confirm, handle_channel_open_init_and_try, Client, IbcChannel,
+    IbcConnections,
+};
+use ckb_ics_axon::message::{
+    Envelope, MsgChannelOpenAck, MsgChannelOpenConfirm, MsgChannelOpenInit, MsgChannelOpenTry,
+    MsgType,
+};
+use ckb_standalone_types::prelude::Entity;
+#[cfg(feature = "debugging")]
+use ckb_std::ckb_types::prelude::*;
+use ckb_std::error::SysError;
+use ckb_std::{ckb_constants::Source, high_level as hl};
+use rlp::decode;
+use tiny_keccak::{Hasher as _, Keccak};
+
+pub fn main() -> Result<()> {
+    let envelope = load_envelope()?;
+    match envelope.msg_type {
+        MsgType::MsgChannelOpenInit
+        | MsgType::MsgChannelOpenTry
+        | MsgType::MsgChannelOpenAck
+        | MsgType::MsgChannelOpenConfirm => {}
+        _ => return Ok(()),
+    }
+
+    let client_data = hl::load_cell_data(0, Source::CellDep).map_err(|_| Error::LoadCellDataErr)?;
+    let client = decode::<Client>(&client_data).map_err(|e| Error::Encoding)?;
+
+    let input_data = hl::load_cell_data(0, Source::GroupInput)?;
+    let expected_input_hash: [u8; 32] = input_data
+        .try_into()
+        .map_err(|e| Error::ChannelHashUnmatch)?;
+
+    let output_data = hl::load_cell_data(0, Source::GroupOutput)?;
+    let expected_output_hash: [u8; 32] = output_data
+        .try_into()
+        .map_err(|e| Error::ChannelHashUnmatch)?;
+
+    let witness_args0 = hl::load_witness_args(0, Source::Input)?;
+    let witness_args1 = hl::load_witness_args(1, Source::Output)?;
+
+    let old_cell_data = witness_args0.input_type();
+    let new_cell_data = witness_args1.output_type();
+
+    if old_cell_data.is_none() || new_cell_data.is_none() {
+        return Err(Error::Encoding);
+    }
+
+    let old_bytes = old_cell_data.to_opt().unwrap();
+    let old_slice = old_bytes.as_slice();
+    if keccak256(old_slice) != expected_input_hash {
+        return Err(Error::ChannelHashUnmatch);
+    }
+
+    let new_bytes = new_cell_data.to_opt().unwrap();
+    let new_slice = new_bytes.as_slice();
+    if keccak256(new_slice) != expected_output_hash {
+        return Err(Error::ChannelHashUnmatch);
+    }
+
+    verify(client, envelope, old_slice, new_slice)
+}
+
+fn verify(client: Client, envelope: Envelope, old_data: &[u8], new_data: &[u8]) -> Result<()> {
+    match &envelope.msg_type {
+        MsgType::MsgChannelOpenInit | MsgType::MsgChannelOpenTry => {
+            let old_connections: IbcConnections = decode(old_data).map_err(|e| Error::Encoding)?;
+            let new_connections: IbcConnections = decode(new_data).map_err(|e| Error::Encoding)?;
+
+            let witness_args3 = hl::load_witness_args(3, Source::Input)?;
+            let channel_cell_data = witness_args3.output_type();
+            if channel_cell_data.is_none() {
+                return Err(Error::Encoding);
```

Comment:
please don't use `Error::Encoing` for responding all of encoding issues, use more specific error codes in detail